provider/azure/environ.go

```diff
@@ -1074,7 +1074,89 @@ func (env *azureEnviron) instances(
 func (env *azureEnviron) AdoptResources(controllerUUID string, fromVersion version.Number) error {
-	return errors.NotImplementedf("AdoptResources")
+	groupClient := resources.GroupsClient{env.resources}
+	resourceClient := resources.Client{env.resources}
+	var failed []string
+
+	apiVersions, err := collectAPIVersions(env.resources)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	resources, err := groupClient.ListResources(env.resourceGroup, "", nil)
+	if err != nil {
+		return errors.Annotate(err, "listing resources")
+	}
+	for resources.Value != nil {
+		for _, resource := range *resources.Value {
+			// We need to set the API version to a value that's
+			// correct for the specific resource type. If we leave it
+			// as the version for the Microsoft.Resources provider we
+			// get NoRegisteredProviderFound errors.
+			resourceClient.APIVersion = apiVersions[*resource.Type]
+			err := env.updateResourceControllerTag(&resourceClient, resource, controllerUUID)
+			if err != nil {
+				logger.Errorf("error updating resource tags for %q: %v", *resource.Name, err)
+				failed = append(failed, *resource.Name)
+			}
+		}
+		resources, err = groupClient.ListResourcesNextResults(resources)
+		if err != nil {
+			return errors.Annotate(err, "getting next page of resources")
+		}
+	}
+
+	if len(failed) > 0 {
+		return errors.Errorf("failed to update controller for some resources: %v", failed)
+	}
+	return nil
+}
+
+func (env *azureEnviron) updateResourceControllerTag(client *resources.Client, stubResource resources.GenericResource, controllerUUID string) error {
+	stubTags := *stubResource.Tags
+	if *stubTags[tags.JujuController] == controllerUUID {
+		// No update needed.
+		return nil
+	}
+
+	namespace, parentPath, subtype, err := splitResourceType(*stubResource.Type)
+	if err != nil {
+		return errors.Annotatef(err, "splitting resource type")
+	}
+
+	// Need to get the resource individually to ensure that the
+	// properties are populated.
+	resource, err := client.Get(env.resourceGroup, namespace, parentPath, subtype, *stubResource.Name)
+	if err != nil {
+		return errors.Annotatef(err, "getting full resource %q", *stubResource.Name)
+	}
+
+	logger.Debugf("updating %s (%s) juju controller uuid to %s",
+		*resource.Name, *resource.Type, controllerUUID)
+	(*resource.Tags)[tags.JujuController] = &controllerUUID
+	_, err = client.CreateOrUpdate(
+		env.resourceGroup,
+		namespace,
+		parentPath,
+		subtype,
+		*resource.Name,
+		resource,
+	)
+	return errors.Annotatef(err, "updating controller for %q", *resource.Name)
+}
+
+// splitResourceType breaks the resource type into provider namespace,
+// parent path and subtype so we can pass the components to the
+// resource CreateOrUpdate method.
+func splitResourceType(resourceType string) (string, string, string, error) {
+	parts := strings.Split(resourceType, "/")
+	if len(parts) < 2 {
+		return "", "", "", errors.Errorf("expected at least 2 parts in resource type %q", resourceType)
```

Comment:
Sometimes (fairly often) the type looks like `Namespace/type`, eg `Microsoft.Compute/virtualMachines` - in that case there's no parent path.